### PR TITLE
[OSCD] T1016 - linux/macOS firewall enumeration

### PR DESCRIPTION
**Pull request by remotephone**
Covering:

MacOs Task 1 : T1016: System Network Configuration Discovery 
Linux Task 2: T1016: System Network Configuration Discovery
(may overlap with https://github.com/Neo23x0/sigma/pull/1084) 

Regarding the outcome of #1048 we can consider renaming the file name prefix and also move it into the right folder.

### CODE REVIEW

File: rules/linux/lnx_system_net_disc_firewall_enum.yml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+id: 71da9e5a-fb1e-46a8-abc1-28c80173af4c
+status: experimental
+description: Detects enumeration of firewall configuration
+author: remotephone, oscd.community
+date: 2020/10/06
+references:
+    - https://github.com/redcanaryco/atomic-red-team/blob/master/atomics/T1016/T1016.md
+    - https://attack.mitre.org/techniques/T1016
+logsource:
+    product: unix
```
Comment (yugoslavskiy) [inline rules/linux/lnx_system_net_disc_firewall_enum.yml:11]: I would rather use macOS category! (your rule is going to be the first one! congratulations!)
this way you better create multiple rules for different platforms.

File: rules/linux/lnx_system_net_disc_firewall_enum.yml [hunk 2]
```diff
@@ -0,0 +1,31 @@
+title: System Network Discovery - Firewall Enumeration
+id: 71da9e5a-fb1e-46a8-abc1-28c80173af4c
+status: experimental
+description: Detects enumeration of firewall configuration
+author: remotephone, oscd.community
+date: 2020/10/06
+references:
+    - https://github.com/redcanaryco/atomic-red-team/blob/master/atomics/T1016/T1016.md
+    - https://attack.mitre.org/techniques/T1016
```
Comment (yugoslavskiy) [inline rules/linux/lnx_system_net_disc_firewall_enum.yml:9]: could you please remove a link to the ATT&CK, as it was already mentioned in `tags`?

File: rules/linux/lnx_system_net_disc_firewall_enum.yml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        - 'arp -a'
+        - 'ip'
+        - 'ss'
+        # macOS and Linux
+        - 'netstat'
+        - 'ifconfig'
+        # macOS only
+        - 'defaults read /Library/Preferences/com.apple.alf'
+        - 'socketfilterfw'
+    condition: keywords
```
Comment (yugoslavskiy) [inline rules/linux/lnx_system_net_disc_firewall_enum.yml:24]: I would recommend you use "commandline" and "ProcessName" here, just like we do it for windows events.

Comment (remotephone) [reply rules/linux/lnx_system_net_disc_firewall_enum.yml]: thanks for the guidance, I have incorporated your recommendations into the latest commit.

File: rules/linux/lnx_system_net_disc_firewall_enum.yml (hunk 1 above)
Comment (remotephone) [reply rules/linux/lnx_system_net_disc_firewall_enum.yml:11]: Done, thank you! These are the first rules I've written in sigma so that's exciting!

Comment (yugoslavskiy) [reply rules/linux/lnx_system_net_disc_firewall_enum.yml:11]: you did a great job, @remotephone!

File: rules/linux/macos_system_net_disc_firewall_enum.yml [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    - https://github.com/redcanaryco/atomic-red-team/blob/master/atomics/T1016/T1016.md
+logsource:
+    product: macos
+detection:
+    selection:
+        ParentCommandLine|contains:
+            - 'netstat'
+            - 'ifconfig'
+            - 'defaults read /Library/Preferences/com.apple.alf'
+            - 'socketfilterfw'
```
Comment (yugoslavskiy) [inline rules/linux/macos_system_net_disc_firewall_enum.yml:17]: could you please change this part to processname/commandline? 

Here is how `netstat` execution looks like in Santa log:

```
 "process": {
      "args": [
        "netstat"
      ],
      "executable": "/usr/sbin/netstat",
    },
```

here is how ifconfig execution looks like:

```
"process": {
      "args": [
        "/sbin/ifconfig"
      ],
      "executable": "/sbin/ifconfig"
    },
```

here is how `defaults read /Library/Preferences/com.apple.alf` execution looks like:

```
"process": {
      "args": [
        "defaults",
        "read",
        "/Library/Preferences/com.apple.alf"
      ],
      "executable": "/usr/bin/defaults"
    },
```

Here is how `/usr/libexec/ApplicationFirewall/socketfilterfw --getglobalstate` execution looks like:

```
"process": {
      "args": [
        "/usr/libexec/ApplicationFirewall/socketfilterfw",
        "--getglobalstate"
      ],
      "executable": "/usr/libexec/ApplicationFirewall/socketfilterfw",
    },
```

So the idea is to split processname and commandline where it is possible. it is cheaper from processing point of view.

File: rules/linux/macos_system_net_disc_firewall_enum.yml [hunk 5]
```diff
@@ -0,0 +1,24 @@
+title: System Network Discovery - Firewall Enumeration
+id: 71da9e5a-fb1e-46a8-abc1-28c80173af4c
+status: experimental
+description: Detects enumeration of firewall configuration
+author: remotephone, oscd.community
+date: 2020/10/06
+references:
+    - https://github.com/redcanaryco/atomic-red-team/blob/master/atomics/T1016/T1016.md
+logsource:
+    product: macos
```
Comment (yugoslavskiy) [inline rules/linux/macos_system_net_disc_firewall_enum.yml:10]: could you please add `category: process_creation` here?

File: rules/linux/lnx_system_net_disc_firewall_enum.yml [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+references:
+    - https://github.com/redcanaryco/atomic-red-team/blob/master/atomics/T1016/T1016.md
+logsource:
+    category: process_creation
+    product: unix
+detection:
+    selection:
+        CommandLine|contains:
+            # Linux Only
+            - 'arp -a'
```
Comment (yugoslavskiy) [inline rules/linux/lnx_system_net_disc_firewall_enum.yml:16]: could you please tell me how arp table reflects firewall configuration?

File: rules/linux/lnx_system_net_disc_firewall_enum.yml [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+            - 'ip'
+            - 'ss'
+            # macOS and Linux
+            - 'netstat'
+            - 'ifconfig'
```
Comment (yugoslavskiy) [inline rules/linux/lnx_system_net_disc_firewall_enum.yml:21]: could you please apply the same approach with processname/commandlie as in macos rule?

File: rules/linux/lnx_system_net_disc_firewall_enum.yml (hunk 6 above)
Comment (remotephone) [reply rules/linux/lnx_system_net_disc_firewall_enum.yml:16]: It does not, I'll be removing that in the next commit. Thanks!

Comment (remotephone) [reply rules/linux/lnx_system_net_disc_firewall_enum.yml:16]: On further review, I think I had tunnel vision focusing on firewall enumeration. 

T1016 is more generally network configuration enumeration, not limited to the firewall. So in this case, I think I should rename the rule to drop "Firewall Enumeration" and simply catch as many network discovery commands as reasonable. The coming commit should reflect that.

File: rules/linux/lnx_system_net_disc_firewall_enum.yml (hunk 7 above)
Comment (remotephone) [reply rules/linux/lnx_system_net_disc_firewall_enum.yml]: This has been addressed in the latest commit on both rules. Thanks!

**Comment by remotephone** [comment]
I have rewritten the rules and reconsidered the approach to them. On further review, I think I had tunnel vision focusing on firewall enumeration.

T1016 is more generally network configuration enumeration, not limited to the firewall. So in this case, I think I should rename the rule to drop "Firewall Enumeration" and simply catch as many network discovery commands as reasonable. 

The rules now cover a broader set of linux and macos network commandline utilities, including platform distribution tools (eg iptables) and distribution specific (eg ufw, firewall-cmd). I imagine there are likely more that could be added, but this should be most of the more popular ones.

File: rules/linux/lnx_system_net_disc_firewall_enum.yml (hunk 3 above)
Comment (yugoslavskiy) [reply rules/linux/lnx_system_net_disc_firewall_enum.yml]: Thank you!

File: rules/linux/lnx_system_net_disc_firewall_enum.yml (hunk 7 above)
Comment (yugoslavskiy) [reply rules/linux/lnx_system_net_disc_firewall_enum.yml]: Thank you!

File: rules/linux/macos_system_network_discovery.yml [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+        ProcessName:
+            - '/usr/sbin/netstat'
+            - '/sbin/ifconfig'
+            - '/usr/sbin/ipconfig'
+            - '/usr/libexec/ApplicationFirewall/socketfilterfw'
+            - '/usr/sbin/networksetup'
+            - '/usr/sbin/arp'
+    selection2:
+        ProcessName: '/usr/bin/defaults'
+        Commandline|contains: 'read /Library/Preferences/com.apple.alf'
```
Comment (yugoslavskiy) [inline rules/linux/macos_system_network_discovery.yml:23]: could you please implement the last fix?
the thing is that both osquery and santa split arguments into a list of strings, without spaces. 
at the same time, having space in such queries makes the detection easily bypassed (just add one more space and that't it).

so I would like to ask you to change it to the following:

```
        Commandline|contains|all: 
            - 'read'
            - '/Library/Preferences/com.apple.alf'
```

The rest of the rule looks great and ready for merge. Thank you!

Comment (remotephone) [reply rules/linux/macos_system_network_discovery.yml]: The latest commit incorporates these changes. Thanks for the explanation, that really helps clarify how I'm thinking about writing these rules!

**Comment by remotephone** [comment]
I can't tell why the checks are failing now. It doesn't appear to be anything in the rules I committed, but maybe I'm missing something?

**Comment by yugoslavskiy** [comment]
@remotephone the tests are broken. I don't see any issues with your rules (: Thank you for your contribution! Well done!

**Review by yugoslavskiy — APPROVED** [review]
@thomaspatzke @Neo23x0 guys, I believe that this PR could be merged (:
please let us know what do you think about it if there is anything you would like to change.

File: rules/linux/lnx_system_network_discovery.yml [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+detection:
+    selection:
+        ProcessName:
+            - '/usr/bin/firewall-cmd'
+            - '/usr/sbin/ufw'
+            - '/usr/sbin/iptables'
+            - '/usr/bin/netstat'
+            - '/usr/bin/ss'
+            - '/usr/sbin/ip'
+            - '/usr/sbin/ifconfig'
```
Comment (yugoslavskiy) [inline rules/linux/lnx_system_network_discovery.yml:21]: sorry for getting to it after the approval. 
I just recalled that paths to the executables could vary from one distribution to another.
I've just checked path to the `ss` on an Ubuntu machine, and it was `/bin/ss`.

So it makes sense to apply the following changes:

```
        ProcessName|endswith:
            - '/firewall-cmd'
            - '/ufw'
            - '/iptables'
            - '/netstat'
            - '/ss'
            - '/ip'
            - '/ifconfig'
```

Could you also please adjust the rest of your Linux rules following this scenario?
Sorry for the inconvenience.

Comment (remotephone) [reply rules/linux/lnx_system_network_discovery.yml]: No worries, thanks for taking time to review this work. Please see the updated commit, it should include your recommendations.

**Comment by remotephone** [comment]
The latest commit should include the changes recommended above and pull [1084](https://github.com/Neo23x0/sigma/pull/1084)

File: rules/linux/lnx_system_network_discovery.yml (hunk 9 above)
Comment (yugoslavskiy) [reply rules/linux/lnx_system_network_discovery.yml]: Thank you very much!

**Comment by remotephone** [comment]
Hi @omergunal, would you like to see any changes in this rule to address the comment [here](https://github.com/Neo23x0/sigma/pull/1084#issuecomment-715650167)?

**Comment by omergunal** [comment]
> Hi @omergunal, would you like to see any changes in this rule to address the comment [here](https://github.com/Neo23x0/sigma/pull/1084#issuecomment-715650167)?

Its ok for me

**Pull request merged**